factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/catchup/paged/PgPagedCatchup.java

```diff
@@ -50,60 +53,63 @@
-            UUID factId = f.id();
-            if (postQueryMatcher.test(f)) {
-              try {
-                subscription.notifyElement(f);
-                log.trace("{} notifyElement called with id={}", request, factId);
-              } catch (MissingTransformationInformation | TransformationException e) {
-                log.warn("{} transformation error: {}", request, e.getMessage());
-                subscription.notifyError(e);
-                throw e;
-              } catch (Throwable e) {
-                // debug level, because it happens regularly
-                // on
-                // disconnecting clients.
-                log.debug("{} exception from subscription: {}", request, e.getMessage());
+
+    SingleConnectionDataSource ds = new SingleConnectionDataSource(connectionSupplier.get(), true);
+    try {
+      val jdbc = new JdbcTemplate(ds);
+
+      jdbc.execute("CREATE TEMPORARY TABLE catchup(ser bigint primary key)");
+      jdbc.execute("CREATE INDEX catchup_tmp_idx1 ON catchup(ser)"); // improves perf on sorting?
+
+      PgCatchUpPrepare prep = new PgCatchUpPrepare(jdbc, request);
+      val numberOfFactsToCatchUp = prep.prepareCatchup(serial);
+      if (numberOfFactsToCatchUp > 0) {
+        try {
+          PgCatchUpFetchPage fetch = new PgCatchUpFetchPage(jdbc, props.getPageSize(), request);
+          while (true) {
+            List<Fact> facts = fetch.fetchFacts(serial);
+
+            for (Fact f : facts) {
+              UUID factId = f.id();
+              if (postQueryMatcher.test(f)) {
                 try {
-                  subscription.close();
-                } catch (Exception e1) {
-                  log.warn("{} exception while closing subscription: {}", request, e1.getMessage());
+                  subscription.notifyElement(f);
+                } catch (MissingTransformationInformation | TransformationException e) {
+                  log.warn("{} transformation error: {}", request, e.getMessage());
+                  subscription.notifyError(e);
+                  throw e;
+                } catch (Throwable e) {
+                  // debug level, because it happens regularly
+                  // on
+                  // disconnecting clients.
+                  log.debug("{} exception from subscription: {}", request, e.getMessage());
+                  try {
+                    subscription.close();
+                  } catch (Exception e1) {
+                    log.warn(
+                        "{} exception while closing subscription: {}", request, e1.getMessage());
+                  }
+                  throw e;
                 }
-                throw e;
+              } else {
+                log.trace("{} filtered id={}", request, factId);
               }
-            } else {
-              log.trace("{} filtered id={}", request, factId);
+            }
+
+            if (facts.isEmpty()) {
+              // we have reached the end
+              break;
```

Comment:
I think a do-while loop instead is more idiomatic

Reply:
true